components/scheduler/executors.go

```diff
@@ -48,16 +67,71 @@ type (
 	}
 )
 
-func (e taskExecutor) executeScheduleTask(
+func (e taskExecutor) executeSchedulerWaitTask(
 	env hsm.Environment,
 	node *hsm.Node,
-	task ScheduleTask,
+	task SchedulerWaitTask,
 ) error {
 	if err := node.CheckRunning(); err != nil {
 		return err
 	}
-	// TODO(Tianyu): Perform scheduler logic before scheduling self again
-	return hsm.MachineTransition(node, func(scheduler Scheduler) (hsm.TransitionOutput, error) {
+	return hsm.MachineTransition(node, func(scheduler *Scheduler) (hsm.TransitionOutput, error) {
 		return TransitionSchedulerActivate.Apply(scheduler, EventSchedulerActivate{})
 	})
 }
+
+func (e taskExecutor) executeSchedulerRunTask(
+	ctx context.Context,
+	env hsm.Environment,
+	ref hsm.Ref,
+	task SchedulerRunTask,
+) error {
+	return env.Access(ctx, ref, hsm.AccessWrite, func(node *hsm.Node) error {
+		if err := node.CheckRunning(); err != nil {
+			return err
+		}
+		s, err := hsm.MachineData[*Scheduler](node)
+		if err != nil {
+			return err
+		}
+
+		s.populateTransientFieldsIfAbsent(e.options.logger, e.options.metricsHandler, e.options.frontendClient, e.options.historyClient)
+
+		if s.Args.State.LastProcessedTime == nil {
+			// log these as json since it's more readable than the Go representation
+			specJson, _ := protojson.Marshal(s.Args.Schedule.Spec)
+			policiesJson, _ := protojson.Marshal(s.Args.Schedule.Policies)
+			s.logger.Info("Starting schedule", tag.NewStringTag("spec", string(specJson)), tag.NewStringTag("policies", string(policiesJson)))
+			s.Args.State.LastProcessedTime = timestamppb.Now()
+			s.Args.State.ConflictToken = scheduler.InitialConflictToken
+			s.Args.Info.CreateTime = s.Args.State.LastProcessedTime
+		}
+
+		t1 := timestamp.TimeValue(s.Args.State.LastProcessedTime)
+		t2 := time.Now()
+		if t2.Before(t1) {
+			// Time went backwards. Currently this can only happen across a continue-as-new boundary.
```

Comment:
This comment is no longer accurate if you're using `time.Now()`. If we don't have a monotonic clock to use than this logic is fine.